### PR TITLE
repo: define gh repository behavior

### PR DESCRIPTION
**Pull request by swaptr**
Designing the repository.

### CODE REVIEW

File: reckless_github/src/repository.rs [hunk 1]
```diff
@@ -35,6 +35,7 @@ impl Repository for Github {
     /// Where to store the index is an implementation
     /// details.
     async fn init(&self) -> Result<(), RecklessError> {
+        println!("{} was initialized", self.name);
```
Comment (swaptr) [inline reckless_github/src/repository.rs:38]: @vincenzopalazzo Why does this not work? As I understand we implement the `Repository` trait, this should mean this custom behavior should work when we call `init()` from the test in `lib.rs`.

File: reckless_github/src/repository.rs [hunk 2]
```diff
@@ -35,6 +37,10 @@ impl Repository for Github {
     /// Where to store the index is an implementation
     /// details.
     async fn init(&self) -> Result<(), RecklessError> {
+        let repo = match git2::Repository::clone(&self.url, get_dir_path_from_url(&self.url)) {
+            Ok(repo) => repo,
+            Err(e) => panic!("failed to open: {}", e),
```
Comment (vincenzopalazzo) [inline reckless_github/src/repository.rs:42]:
```suggestion
            Err(e) => return RecklessError::new(format!("failed to open: {}"));,
```

Or something like that I do not know if this compile, do not panic, but return an error instead

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]

File: reckless_github/src/repository.rs [hunk 3]
```diff
@@ -35,6 +37,13 @@ impl Repository for Github {
     /// Where to store the index is an implementation
     /// details.
     async fn init(&self) -> Result<(), RecklessError> {
+        let repo =
+            match git2::Repository::clone_recurse(&self.url, get_dir_path_from_url(&self.url)) {
+                Ok(repo) => repo,
+                Err(e) => panic!("failed to open: {}", e),
+            };
```
Comment (vincenzopalazzo) [inline reckless_github/src/repository.rs:44]: I think this is very bugged and I will contact the team to understand why, for now this is a workaround

```diff
diff --git a/reckless_github/src/repository.rs b/reckless_github/src/repository.rs
index c95e403..0f219ef 100644
--- a/reckless_github/src/repository.rs
+++ b/reckless_github/src/repository.rs
@@ -37,14 +37,27 @@ impl Repository for Github {
     /// Where to store the index is an implementation
     /// details.
     async fn init(&self) -> Result<(), RecklessError> {
-        let repo =
-            match git2::Repository::clone_recurse(&self.url, get_dir_path_from_url(&self.url)) {
-                Ok(repo) => repo,
-                Err(e) => panic!("failed to open: {}", e),
-            };
+        let res = git2::Repository::clone(&self.url, get_dir_path_from_url(&self.url));
 
-        println!("{} was initialized", self.name);
-        Ok(())
+        match res {
+            Ok(repo) => {
+                // FIXME: put this inside a utils methods and call the method `clone_recursive_fix`.
+                for sub in repo.submodules().unwrap_or_default() {
+                    println!("URL {}", sub.url().unwrap());
+                    let path = format!(
+                        "{}/{}",
+                        get_dir_path_from_url(&self.url),
+                        sub.path().to_str().unwrap()
+                    );
+                    match git2::Repository::clone(sub.url().unwrap(), path) {
+                        Ok(_) => println!("Ok {:?}", sub.path()),
+                        Err(err) => println!("KO {:?} {:?}", sub.path(), err),
+                    }
+                }
+                Ok(())
+            }
+            Err(err) => Err(RecklessError::new(1, err.message())),
+        }
     }

```

Also with some rust improvement 😄

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]

File: reckless_lib/src/utils.rs [hunk 4]
```diff
@@ -19,17 +20,37 @@ pub fn create_dir_in_home(relative_path: &str) {
...
         .unwrap();
-    return path + "/.reckless/" + &url.split(".com/").last().unwrap().trim().to_string();
+    return format!(
+        "{}/.reckless{}",
+        path,
+        &url.split(".com").last().unwrap().trim().to_string()
+    );
+}
+
+pub fn clone_recursive_fix(repo: git2::Repository, url: &str) {
```
Comment (vincenzopalazzo) [inline reckless_lib/src/utils.rs:35]:
```suggestion
pub fn clone_recursive_fix(repo: git2::Repository, url: &str) -> Result<(), RecklessError> {
```

File: reckless_lib/src/utils.rs [hunk 5]
```diff
@@ -19,17 +20,37 @@ pub fn create_dir_in_home(relative_path: &str) {
...
+    for sub in repo.submodules().unwrap_or_default() {
+        println!("URL {}", sub.url().unwrap());
+        let path = format!(
+            "{}/{}",
+            get_dir_path_from_url(url),
+            sub.path().to_str().unwrap()
+        );
+        match git2::Repository::clone(sub.url().unwrap(), path) {
+            // Fix error handling
+            Ok(_) => println!("Ok {:?}", sub.path()),
```
Comment (vincenzopalazzo) [inline reckless_lib/src/utils.rs:45]:
```suggestion
            Ok(_) => Ok(()),
```

File: reckless_lib/src/utils.rs [hunk 6]
```diff
@@ -19,17 +20,37 @@ pub fn create_dir_in_home(relative_path: &str) {
...
+            Err(err) => println!("KO {:?} {:?}", sub.path(), err),
```
Comment (vincenzopalazzo) [inline reckless_lib/src/utils.rs:46]:
```suggestion
            Err(err) => Err(RecklessError::new(1, err.message())),
```

File: reckless_github/src/repository.rs [hunk 7]
```diff
@@ -37,14 +37,14 @@ impl Repository for Github {
...
-                Err(e) => panic!("failed to open: {}", e),
-            };
-
-        println!("{} was initialized", self.name);
-        Ok(())
+        let res = git2::Repository::clone(&self.url, get_dir_path_from_url(&self.url));
+        match res {
+            Ok(repo) => {
+                clone_recursive_fix(repo, &self.url);
+                Ok(())
```
Comment (vincenzopalazzo) [inline reckless_github/src/repository.rs:44]:
```suggestion
                clone_recursive_fix(repo, &self.url)
```

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]
Remember to never abort but return an error inside an API, when you use the API you can abort, but not inside the API

File: reckless_lib/src/utils.rs [hunk 8]
```diff
@@ -1,4 +1,5 @@
-use std::{env, fs::create_dir_all, path::Path};
+use git2;
+use std::{env, fmt::format, fs::create_dir_all, path::Path};
```
Comment (vincenzopalazzo) [inline reckless_lib/src/utils.rs:2]:
```suggestion
use std::fmt::format;
use std::env;
use std::fs::create_dir_all;
use std::path::Path;
```

It is not a rule, it is just my preference

File: reckless_lib/src/utils.rs [hunk 9]
```diff
@@ -19,17 +20,37 @@ pub fn create_dir_in_home(relative_path: &str) {
...
+    return format!(
+        "{}/.reckless{}",
+        path,
+        &url.split(".com").last().unwrap().trim().to_string()
+    );
+}
+
+pub fn clone_recursive_fix(repo: git2::Repository, url: &str) {
+    for sub in repo.submodules().unwrap_or_default() {
+        println!("URL {}", sub.url().unwrap());
```
Comment (vincenzopalazzo) [inline reckless_lib/src/utils.rs:37]:
```suggestion
        println!("URL {}", sub.url().unwrap());
```

we should use a log library, but not in this PR

File: reckless_lib/src/utils.rs (hunk 4 above)
Comment (swaptr) [reply reckless_lib/src/utils.rs]: Yes, that would have been my general idea. I will keep this is mind.

File: reckless_lib/src/utils.rs (hunk 8 above)
Comment (swaptr) [reply reckless_lib/src/utils.rs]: Done.

**Review by vincenzopalazzo — APPROVED** [review]
Awesome work, thanks!

**Pull request merged**